### ISSUE TITLE
Feature: Support ?expands query arg

### ISSUE THREAD

**Issue by boydgreenfield** [biosustain/potion#62]
It looks like this has already been discussed in https://github.com/biosustain/potion/issues/27, but I think it would be very useful to support an `?expands` query param for expanding Potion JSON responses. 

Stripe, among others, do this in their API (though using `-d` to curl?) – https://stripe.com/docs/api/curl#expanding_objects

As an initial proof-of-concept, this commit implements a quite simple, 1-level expansion scheme*: https://github.com/biosustain/potion/commit/521e39c905a5d40fb48e4a192077683711f4278f

The above allows one to pass `?expand=all` or `?expand=<resource_name>` or `?expand=<resource1>&expand=<resource2>` etc. It also includes adding an `expand` attribute on `Resource.meta` that auto-expands everything in the resource. 

A few open design and implementation questions:

1. On what links/routes is this useful, where's the right place to insert it into the relevant schemas, and where do we want to parse the Flask `request` argument (if not in `fields.py` directly)?
2. Do we want a Resource-level `expand` option for auto-expanding all `ToOne` and `ToMany` reference fields? If so, should we check for an `?expand=false` as an alternate argument? 
3. Instead of 2, do we want to have this as an argument to the `ToOne` and `ToMany` field constructors? This would allow field-level automatic expansion as warranted. If we do this, maybe we can collapse everything with `?expand=false` or use a `?collapse` param to toggle them off individually (though that gets more complex)

Happy to draft up a PR that at least implements the field-level checking depending on the answers to 2 and 3 above. That said, I am less familiar with the schema generation and how'd you'd like to structure Flask `request` object parsing, so happy to take either guidance or follow-on commits there.

* For a variety of reasons, deeper than 1-level expansion, seems like a Bad Idea™. Here I'm preventing recursion by just setting the name of the resource on the Flask request global and then checking if `field.resource` matches (and is thus top-level). There are probably other decent ways of doing this by passing an `expands` kwarg through... but I opted for this approach after seeing how many function signatures it looked like the latter would require.

**Comment by lyschoening** [comment]
It goes a bit against the spirit of Potion. The ideal solution would be to use HTTP/2 server push. However, it's not likely that Flask will support that any time soon. (I am planning a version of Potion not based on Flask that might have this feature, but that would of course not be viable for someone using additional Flask extensions.)

With bare HTTP/2 there is then twice the latency (1st request getting the resource, 2nd round of requests getting all the references that are not yet cached). There may be a few scenarios where that is still not fast enough though, so I am willing to consider an 'expand' feature.

The expand option, if implemented, should be part of the `Instances` logic so that it is documented in the schema for the GET request. `Resource.read` should probably have that argument too. `FieldSet.format()` could have an additional 'expand' argument that would take a list of expandable attributes. 

Regarding the configuration options, I guess there needs to be at least (1) a way to specify which fields are expandable (all/none/some) and which of these are expanded by default (all/none/some). We can either define this in `Meta` or on the `ToOne`/`ToMany` fields themselves. If we define it on `Meta`, the `FieldSet` class would generate an appropriate schema and take care of the formatting. Otherwise, `ToX()` would then take some more keyword arguments and `ToX.format()` would get an expand argument. I'm tending towards having all the logic in `FieldSet` and the configuration in `Meta` to keep the field classes simple. 

There are a few concerns that will no doubt pop up. For instance with Principals a resource could in theory have a reference to an item the user does not have read access to. If that reference is made expandable, the user will be able to read it regardless of their permissions. There could either be a check, during initialization, whether all included expandable fields have read permissions that are a subset of the parent resource, or the read permissions could be dynamically updated to include those of the expanded resource types. The alternative is to leave that sanity check up to the developer.

**Comment by boydgreenfield** [comment]
I agree that it's somewhat inelegant, but I do think the practical utility of an `?expand` param is very significant – the specific case where I'm finding we need this is on `Instances` calls, where you have a few embedded relationships (e.g., metadata on an object that you don't need for all use cases, but do for some). 

Then your 1 initial response becomes 1 + `page_size * number_of_relationships` additional secondary queries – which in practice is also much more than 2x latency 

Agreed re: the nuances on Principals – seems like it's probably tractable to put some protection in via init checks, but worth having that quite explicit so folks don't inadvertently expose relationships. So I suppose that's another argument in favor of a field-level setting: Having it be default False and then only enabled in the case(s) the developer chooses. (A second argument for this is that one has to specify the relationship fields in the schema manually _anyway_, so it's not to onerous to specify it there vs. in a `Resource.meta` option)

**Comment by upmauro** [comment]
+1
